packages/grid/x-data-grid-pro/src/hooks/features/columnResize/useGridColumnResize.tsx

```diff
@@ -426,12 +511,164 @@ export const useGridColumnResize = (
+
+      colDefRef.current = colDef as GridStateColDef;
+      colElementRef.current =
+        apiRef.current.columnHeadersContainerElementRef?.current!.querySelector<HTMLDivElement>(
+          `[data-field="${colDef.field}"]`,
+        )!;
+
+      const headerFilterRowElement = apiRef.current.headerFiltersElementRef?.current;
+
+      if (headerFilterRowElement) {
+        headerFilterElementRef.current = headerFilterRowElement.querySelector<HTMLDivElement>(
+          `[data-field="${colDef.field}"]`,
+        ) as HTMLDivElement;
+      }
+
+      colGroupingElementRef.current = findGroupHeaderElementsFromField(
+        apiRef.current.columnHeadersContainerElementRef?.current!,
+        colDef.field,
+      );
+
+      colCellElementsRef.current = findGridCellElementsFromCol(
+        colElementRef.current,
+        apiRef.current,
+      );
+
+      const doc = ownerDocument(apiRef.current.rootElementRef!.current);
+      doc.body.style.cursor = 'col-resize';
+
+      resizeDirection.current = getResizeDirection(event.currentTarget, theme.direction);
+
+      initialOffsetToSeparator.current = computeOffsetToSeparator(
+        event.clientX,
+        colElementRef.current!.getBoundingClientRect(),
+        resizeDirection.current,
+      );
+
+      doc.addEventListener('mousemove', handleResizeMouseMove);
+      doc.addEventListener('mouseup', handleResizeMouseUp);
+
+      // Prevent the click event if we have resized the column.
+      // Fixes https://github.com/mui/mui-x/issues/4777
+      doc.addEventListener('click', preventClick, true);
+    });
+
+  const handleColumnSeparatorDoubleClick: GridEventListener<'columnSeparatorDoubleClick'> =
+    useEventCallback((_, event) => {
+      if (props.disableAutosize) {
+        return;
+      }
+
+      // Only handle left clicks
+      if (event.button !== 0) {
+        return;
+      }
+
+      apiRef.current.autosizeColumns(props.autosizeOptions);
+    });
+
+  /**
+   * API METHODS
+   */
+
+  const columnVirtualizationDisabled = useColumnVirtualizationDisabled(apiRef);
+  const isAutosizingRef = React.useRef(false);
+  const autosizeColumns = React.useCallback<GridColumnResizeApi['autosizeColumns']>(
+    async (userOptions) => {
+      const root = apiRef.current.rootElementRef?.current;
+      if (!root) {
+        return;
+      }
+      if (isAutosizingRef.current) {
+        return;
+      }
+      isAutosizingRef.current = true;
+
+      const state = gridColumnsStateSelector(apiRef.current.state);
+      const options = {
+        ...DEFAULT_GRID_AUTOSIZE_OPTIONS,
+        ...userOptions,
+        columns: userOptions?.columns ?? state.orderedFields.map((field) => state.lookup[field]),
+      };
+      options.columns = options.columns.filter(
+        (c) => state.columnVisibilityModel[c.field] !== false,
+      );
+
+      try {
+        apiRef.current.unstable_setColumnVirtualization(false);
```

Comment:
I'm not sure about this... It feels slow with 100 columns: https://codesandbox.io/s/wizardly-water-23jrqs?file=/Demo.tsx
Did you consider limiting autosizing to currently rendered columns?

Reply:
By the way, it seems like a double click on column separator resizes all columns. Is it intentional?


https://github.com/mui/mui-x/assets/13808724/f3a59eb9-ccfb-48b5-affb-0bfececd5f1d

Reply:
> Did you consider limiting autosizing to currently rendered columns?

I didn't, I'm pretty sure there are cases where people would want to autosize everything. I can add an option to limit to currently rendered rows.

> By the way, it seems like a double click on column separator resizes all columns. Is it intentional?

Yes, I might have misunderstood how it's implemented elsewhere. I didn't test other grids with more than 1 column so I didn't notice it was applying to only 1 column.

Reply:
Should we add a way for users to autosize all columns? In the toolbar maybe? I feel like it's an useful feature.

Reply:
> Should we add a way for users to autosize all columns?

Yes, what do you think about making it opt-in?

> In the toolbar maybe?

This resonates with the toolbar redesign @gerdadesign explored in https://www.figma.com/file/tOAZwAT1ebcTH4tb4zv3Lo/Pivoting?type=design&node-id=3-16&mode=design&t=GARKMvWXyWreD6Fr-0

<img width="803" src="https://github.com/mui/mui-x/assets/13808724/020cb6ed-7da0-4d93-8e68-cd23f44479ef">

We can add it to the settings menu on the right 👍🏻

Reply:
Do we want to do this after the new toolbar is implemented?

Reply:
I've opened #10371, we can discuss more there, it's not blocking for this PR.

Reply:
Correct me if I am wrong, but If I understand it correctly, the current behaviour is to autosize all the columns (as we are disabling column virtualization during the process), if that's the case, shouldn't the new issue say `Allow users to autosize virtualized columns only` 🤔